From b4cb993d92ac1479e46cee8f57a13048a19fa65f Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Wed, 17 Jun 2020 14:41:45 +0200
Subject: [PATCH 1/1] efi_loader: simplify efi_set_variable_int()

Reduce the code size by simplifying efi_set_variable_int().
The size of u-boot.bin is reduced by 138 bytes for qemu_arm64_defconfig.

The internal representation of UEFI variables will have a trailing comma if
the variable is not authenticated.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_variable.c | 36 +++++++++++------------------------
 1 file changed, 11 insertions(+), 25 deletions(-)

diff --git a/lib/efi_loader/efi_variable.c b/lib/efi_loader/efi_variable.c
index 85df1427bc..907f621a3b 100644
--- a/lib/efi_loader/efi_variable.c
+++ b/lib/efi_loader/efi_variable.c
@@ -964,32 +964,18 @@ efi_status_t efi_set_variable_int(u16 *variable_name, const efi_guid_t *vendor,
 	/*
 	 * store attributes
 	 */
-	attributes &= (EFI_VARIABLE_READ_ONLY |
-		       EFI_VARIABLE_NON_VOLATILE |
-		       EFI_VARIABLE_BOOTSERVICE_ACCESS |
-		       EFI_VARIABLE_RUNTIME_ACCESS |
-		       EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS);
 	s += sprintf(s, "{");
-	while (attributes) {
-		attr = 1 << (ffs(attributes) - 1);
-
-		if (attr == EFI_VARIABLE_READ_ONLY) {
-			s += sprintf(s, "ro");
-		} else if (attr == EFI_VARIABLE_NON_VOLATILE) {
-			s += sprintf(s, "nv");
-		} else if (attr == EFI_VARIABLE_BOOTSERVICE_ACCESS) {
-			s += sprintf(s, "boot");
-		} else if (attr == EFI_VARIABLE_RUNTIME_ACCESS) {
-			s += sprintf(s, "run");
-		} else if (attr ==
-			   EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS) {
-			s += sprintf(s, "time=");
-			s = bin2hex(s, (u8 *)&time, sizeof(time));
-		}
-
-		attributes &= ~attr;
-		if (attributes)
-			s += sprintf(s, ",");
+	if (attributes & EFI_VARIABLE_READ_ONLY)
+		s += sprintf(s, "ro,");
+	if (attributes & EFI_VARIABLE_NON_VOLATILE)
+		s += sprintf(s, "nv,");
+	if (attributes & EFI_VARIABLE_BOOTSERVICE_ACCESS)
+			s += sprintf(s, "boot,");
+	if (attributes &EFI_VARIABLE_RUNTIME_ACCESS)
+		s += sprintf(s, "run,");
+	if (attributes & EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS) {
+		s += sprintf(s, "time=");
+		s = bin2hex(s, (u8 *)&time, sizeof(time));
 	}
 	s += sprintf(s, "}");
 	s += sprintf(s, "(blob)");
-- 
2.27.0

